### ISSUE TITLE
Segfault after "GtkWarning: Invalid text buffer iterator"

### ISSUE THREAD

**Issue by vossenjp** [jaap-karssenberg/zim-desktop-wiki#16]
I'm opening this somewhat as a place-holder while I explore the issue(s).

First, Zim has been rock-solid since I've been using it, until 0.66 (via PPA), when I've had:
```
$ sudo zgrep 'zim' /var/log/syslog*
/var/log/syslog:May  2 10:49:35 weber4 kernel: [409833.768734] zim[17915]: segfault at 24 ip 00007f4e6aae677b sp 00007f4e53ffda60 error 4 in libgtk-x11-2.0.so.0.2400.23[7f4e6a91c000+430000]
/var/log/syslog.1:May  1 15:15:51 weber4 kernel: [339753.406342] zim[27419]: segfault at 24 ip 00007fef308e977b sp 00007fef1e8e74d0 error 4 in libgtk-x11-2.0.so.0.2400.23[7fef3071f000+430000]
/var/log/syslog.1:May  1 15:17:09 weber4 kernel: [339831.194160] zim[27430]: segfault at 24 ip 00007f7a454e177b sp 00007f7a34c1baa0 error 4 in libgtk-x11-2.0.so.0.2400.23[7f7a45317000+430000]
/var/log/syslog.1:May  1 15:20:30 weber4 kernel: [340032.146867] zim[27449]: segfault at 24 ip 00007f553e98977b sp 00007f552e0ad4d0 error 4 in libgtk-x11-2.0.so.0.2400.23[7f553e7bf000+430000
```
I can't pin down the crashes, other than to say that they only happen when I and typing or pasting.  I'll be in the middle of adding a line and poof, Zim is gone.  If wonder if there might be a render/save/dirty-flag race condition now?

Second, when re-starting after those crashes it has taken 1m20s by stopwatch from clicking on the fully up-to-date Mint-17.3 Cinnamon panel icon to get to a rendered window.  This is in a VirtualBox VM on a Windows8 host (sigh, $WORK T460 laptop) on a Bitlocker encrypted SSD.  Under normal/previous circumstances Zim launches almost instantly, certainly in < 3-4 seconds.

I'm not at all sure if the 2 issues are really related, more time/testing needed.

**Comment by jaap-karssenberg** [comment]
Maybe related to same root case as issue #7 which is also not understood at this time

**Comment by vossenjp** [comment]
I may have captured useful output from a segfault, see "### ... then out of nowhere" in the console output below.

I had several more crashes yesterday, during bad times, and it took 1m43s, 2m45s, and 2+mins to restart (very frustrating).  In between one of the crashes I did an `rm -rf .../.zim` which didn't seem to help, but I'll get back to that.  I also ran it in a terminal, and annotated output is below.

It seems like if it crashes and you just restart it will crash again soon.  But if it crashes and you restart, then gracefully close, then start again...it works for a lot longer.  And/or, that `rm` I did helped but needed a graceful shutdown and restart to really help?

Leading "'###" are my annotations.

```
### 2017-05-02_22:05:59_EDT
$ zim
INFO: This is zim 0.66
WARNING: Could not find all icon sizes for the application icon
INFO: Using default notebook: file:///home/jp/.../Zim
INFO: Open page: 1 Work:1 Con Calls (1 Work:1 Con Calls)
INFO: Open page: 1 Work:1 Con Calls (1 Work:1 Con Calls)
### After 2m 57s, GUI now open

INFO: Spawning: ['firefox', 'http://.../redmine/projects/.../news/new'] (cwd: None)
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/zim/gui/pathbar.py", line 474, in on_button_clicked
    self.ui.open_page(button.zim_path)
  File "/usr/lib/python2.7/dist-packages/zim/actions.py", line 132, in func
    self.func(instance, *args, **kwargs)
  File "/usr/lib/python2.7/dist-packages/zim/gui/__init__.py", line 728, in open_page
    raise AssertionError, 'Could not close page'
AssertionError: Could not close page
INFO: Open page: 1 Work (1 Work)
INFO: Open page: 1 Work (1 Work)
### VERY SLOW to switch pages, error above was non-issue

Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/zim/signals.py", line 414, in callback
    self.cb_func(*arg, **kwarg)
  File "/usr/lib/python2.7/dist-packages/zim/gui/widgets.py", line 2027, in update_completion
    self._fill_completion_for_anchor(path, prefix, text)
  File "/usr/lib/python2.7/dist-packages/zim/gui/widgets.py", line 2053, in _fill_completion_for_anchor
    for p in self.notebook.pages.list_pages(path):
  File "/usr/lib/python2.7/dist-packages/zim/notebook/index/pages.py", line 485, in list_pages
    page_id = self._pages.get_page_id(path) # can raise
  File "/usr/lib/python2.7/dist-packages/zim/notebook/index/pages.py", line 310, in get_page_id
    raise IndexNotFoundError, 'Page not found in index: %s' % pagename.name
zim.notebook.index.base.IndexNotFoundError: Page not found in index: http
INFO: Open page: 1 Work:1 Con Calls (1 Work:1 Con Calls)
INFO: Open page: 1 Work:1 Con Calls (1 Work:1 Con Calls)
INFO: Open page: 1 Work:1 Con Calls (1 Work:1 Con Calls)
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/zim/signals.py", line 414, in callback
    self.cb_func(*arg, **kwarg)
  File "/usr/lib/python2.7/dist-packages/zim/gui/widgets.py", line 2027, in update_completion
    self._fill_completion_for_anchor(path, prefix, text)
  File "/usr/lib/python2.7/dist-packages/zim/gui/widgets.py", line 2053, in _fill_completion_for_anchor
    for p in self.notebook.pages.list_pages(path):
  File "/usr/lib/python2.7/dist-packages/zim/notebook/index/pages.py", line 485, in list_pages
    page_id = self._pages.get_page_id(path) # can raise
  File "/usr/lib/python2.7/dist-packages/zim/notebook/index/pages.py", line 310, in get_page_id
    raise IndexNotFoundError, 'Page not found in index: %s' % pagename.name
zim.notebook.index.base.IndexNotFoundError: Page not found in index: 1 Work:http
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/zim/signals.py", line 414, in callback
    self.cb_func(*arg, **kwarg)
  File "/usr/lib/python2.7/dist-packages/zim/gui/widgets.py", line 2027, in update_completion
    self._fill_completion_for_anchor(path, prefix, text)
  File "/usr/lib/python2.7/dist-packages/zim/gui/widgets.py", line 2053, in _fill_completion_for_anchor
    for p in self.notebook.pages.list_pages(path):
  File "/usr/lib/python2.7/dist-packages/zim/notebook/index/pages.py", line 485, in list_pages
    page_id = self._pages.get_page_id(path) # can raise
  File "/usr/lib/python2.7/dist-packages/zim/notebook/index/pages.py", line 310, in get_page_id
    raise IndexNotFoundError, 'Page not found in index: %s' % pagename.name
zim.notebook.index.base.IndexNotFoundError: Page not found in index: 1 Work:http
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/zim/signals.py", line 414, in callback
    self.cb_func(*arg, **kwarg)
  File "/usr/lib/python2.7/dist-packages/zim/gui/widgets.py", line 2027, in update_completion
    self._fill_completion_for_anchor(path, prefix, text)
  File "/usr/lib/python2.7/dist-packages/zim/gui/widgets.py", line 2053, in _fill_completion_for_anchor
    for p in self.notebook.pages.list_pages(path):
  File "/usr/lib/python2.7/dist-packages/zim/notebook/index/pages.py", line 485, in list_pages
    page_id = self._pages.get_page_id(path) # can raise
  File "/usr/lib/python2.7/dist-packages/zim/notebook/index/pages.py", line 310, in get_page_id
    raise IndexNotFoundError, 'Page not found in index: %s' % pagename.name
zim.notebook.index.base.IndexNotFoundError: Page not found in index: 1 Work:http
### Don't know what all that is, but no crashes, working fine...

INFO: Open page: 1 Work (1 Work)
INFO: Open page: 1 Work (1 Work)
### MUCH faster node switch

Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/zim/signals.py", line 414, in callback
    self.cb_func(*arg, **kwarg)
  File "/usr/lib/python2.7/dist-packages/zim/gui/widgets.py", line 2027, in update_completion
    self._fill_completion_for_anchor(path, prefix, text)
  File "/usr/lib/python2.7/dist-packages/zim/gui/widgets.py", line 2053, in _fill_completion_for_anchor
    for p in self.notebook.pages.list_pages(path):
  File "/usr/lib/python2.7/dist-packages/zim/notebook/index/pages.py", line 485, in list_pages
    page_id = self._pages.get_page_id(path) # can raise
  File "/usr/lib/python2.7/dist-packages/zim/notebook/index/pages.py", line 310, in get_page_id
    raise IndexNotFoundError, 'Page not found in index: %s' % pagename.name
zim.notebook.index.base.IndexNotFoundError: Page not found in index: http
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/zim/signals.py", line 414, in callback
    self.cb_func(*arg, **kwarg)
  File "/usr/lib/python2.7/dist-packages/zim/gui/widgets.py", line 2027, in update_completion
    self._fill_completion_for_anchor(path, prefix, text)
  File "/usr/lib/python2.7/dist-packages/zim/gui/widgets.py", line 2053, in _fill_completion_for_anchor
    for p in self.notebook.pages.list_pages(path):
  File "/usr/lib/python2.7/dist-packages/zim/notebook/index/pages.py", line 485, in list_pages
    page_id = self._pages.get_page_id(path) # can raise
  File "/usr/lib/python2.7/dist-packages/zim/notebook/index/pages.py", line 310, in get_page_id
    raise IndexNotFoundError, 'Page not found in index: %s' % pagename.name
zim.notebook.index.base.IndexNotFoundError: Page not found in index: http
INFO: Spawning: ['firefox', 'http://.../redmine/issues/591'] (cwd: None)
INFO: Spawning: ['firefox', 'http://.../redmine/news/257'] (cwd: None)
INFO: Open page: 1 Work:1 Con Calls (1 Work:1 Con Calls)
INFO: Open page: 1 Work:1 Con Calls (1 Work:1 Con Calls)
INFO: Open page: 1 Work (1 Work)
INFO: Open page: 1 Work (1 Work)
### So far so good...  2017-05-03 Wed: 12:06 AM EDT, going to bed...

### ...and back to work this morning, working fine, then out of nowhere:
/usr/lib/python2.7/dist-packages/zim/gui/pageview.py:2416: GtkWarning: Invalid text buffer iterator: either the iterator is uninitialized, or the characters/pixbufs/widgets in the buffer have been modified since the iterator was created.
You must use marks, character numbers, or line numbers to preserve a position across buffer modifications.
You can apply tags and insert marks without invalidating your iterators,
but any mutation that affects 'indexable' buffer contents (contents that can be referred to by character offset)
will invalidate all outstanding iterators
  if not bound.is_end() and bound.forward_to_tag_toggle(None):
Segmentation fault

$ grep 'zim' /var/log/syslog | grep -v 'CRON'
May  3 10:14:04 weber4 kernel: [494103.547420] zim[16512]: segfault at 24 ip 00007f8b2052f77b sp 00007f8b0e38ca60 error 4 in libgtk-x11-2.0.so.0.2400.23[7f8b20365000+430000]
```

The "noise" errors may be because I'm using Zim to create notes that end up in Redmine, and I'm using Redmine wiki markup like "link text":http://url and Zim might be trying to interpret them as "Links that start with a ':' are resolved from the top level of the notebook":
* "Page not found in index: 1 Work:http"
* "Page not found in index: http"

I have `grep -c 'http'` = 288 and `grep -c ':http'` = 13 just in 1_Work.txt and I probably have lots of both elsewhere too.

After the crash I restarted Zim, which took a while, then I went to "File > Quit" to close it gracefully since that seems to help and...VBox crashed, taking my entire VM with it.  I don't know that I can blame Zim, since the VM runs on Windows, but the timing was certainly suspicious.

When I restarted the VM, Zim launched in <2 seconds.

Also, in reviewing my notes I *think* that I created `~/.local/share/zim/urls.list` the same day the crashes started...  That file's contents are tab delimited:
```
rmi     http://.../redmine/issues/
rmn     http://.../redmine/news/
jira    http://.../jira/browse/
svn     http://.../viewvc?view=revision&revision=
```

**Comment by vossenjp** [comment]
Zim 0.66 Crashy

2017-05-03 Wed:
* 10:14 AM: **Crashed and killed VBox/VM!**
    * `May  3 10:14:04 weber4 kernel: [494103.547420] zim[16512]: segfault at 24 ip 00007f8b2052f77b sp 00007f8b0e38ca60 error 4 in libgtk-x11-2.0.so.0.2400.23[7f8b20365000+430000]`
* 02:28 PM: Taking FOREVER to switch from "4 Ref" to "1 Work" node
    * Seems hung, but clicking on "4 Ref" then "1 Work" again worked
* 02:51 PM:crashed while typing
    * `May  3 14:51:22 weber4 kernel: [15343.589854] zim[9403]: segfault at 24 ip 00007f5ddbf7a77b sp 00007f5dcb77ca60 error 4 in libgtk-x11-2.0.so.0.2400.23[7f5ddbdb0000+430000]`
    * Relaunch from panel icon took only 24 seconds
    * BUT, it is NOT remembering where I left off!  It's restarting in the wrong node in the wrong place
    * "File > Quit"
    * Launch from panel icon took only 24 seconds, and did start in the right place in the right node
* 07:50 PM: crashed while typing
* 08:11 PM: Attempting to "File > Quit" **CRASHED the VBox VM again!**  Windows error:
```
The instruction at 0x1062a070 reference memory at 0x0000000c.  The memory could not be read.
Click on OK to terminate the program.
```
* 08:18 PM: crashed while typing, JUST AFTER A VM CRASH and reboot!
* 08:21 PM: crashed while typing AGAIN, lost 20-30 characters
* 08:22 PM: re-launch from the CLI, took 40s to rendered GUI
* 08:25 PM: crashed while typing AGAIN, lost maybe 10 characters
```
/usr/lib/python2.7/dist-packages/zim/gui/pageview.py:2334: GtkWarning: Invalid text buffer iterator: either the iterator is uninitialized, or the characters/pixbufs/widgets in the buffer have been modified since the iterator was created.
You must use marks, character numbers, or line numbers to preserve a position across buffer modifications.
You can apply tags and insert marks without invalidating your iterators,
but any mutation that affects 'indexable' buffer contents (contents that can be referred to by character offset)
will invalidate all outstanding iterators
  while iter.compare(end) == -1:
/usr/lib/python2.7/dist-packages/zim/gui/pageview.py:2335: GtkWarning: Invalid text buffer iterator: either the iterator is uninitialized, or the characters/pixbufs/widgets in the buffer have been modified since the iterator was created.
You must use marks, character numbers, or line numbers to preserve a position across buffer modifications.
You can apply tags and insert marks without invalidating your iterators,
but any mutation that affects 'indexable' buffer contents (contents that can be referred to by character offset)
will invalidate all outstanding iterators
  pixbuf = iter.get_pixbuf()
Segmentation fault
```
* 08:26 PM: re-launch from the CLI, took 68s to rendered GUI
* 08:30 PM: crashed while typing AGAIN, lost maybe 14 characters
```
INFO: This is zim 0.66
WARNING: Could not find all icon sizes for the application icon
INFO: Using default notebook: file:///home/jp/webtop/Zim
INFO: Open page: 1 Work:1 Con Calls (1 Work:1 Con Calls)
INFO: Open page: 1 Work:1 Con Calls (1 Work:1 Con Calls)
/usr/lib/python2.7/dist-packages/zim/gui/pageview.py:2706: GtkWarning: Invalid text buffer iterator: either the iterator is uninitialized, or the characters/pixbufs/widgets in the buffer have been modified since the iterator was created.
You must use marks, character numbers, or line numbers to preserve a position across buffer modifications.
You can apply tags and insert marks without invalidating your iterators,
but any mutation that affects 'indexable' buffer contents (contents that can be referred to by character offset)
will invalidate all outstanding iterators
  bound.forward_char()
Segmentation fault
```
* 08:31 PM: re-launch from the CLI, took 68s to rendered GUI
    * Worked fine for a while, then I was away for an hour+
* 10:50 PM: crashed while typing AGAIN, lost 12 characters
```
INFO: This is zim 0.66
WARNING: Could not find all icon sizes for the application icon
INFO: Using default notebook: file:///home/jp/webtop/Zim
INFO: Open page: 1 Work:1 Con Calls (1 Work:1 Con Calls)
INFO: Open page: 1 Work:1 Con Calls (1 Work:1 Con Calls)
INFO: Spawning: ['firefox', 'http://.../redmine/projects/devops/news/new'] (cwd: None)
INFO: Open page: 1 Work (1 Work)
INFO: Open page: 1 Work (1 Work)
/usr/lib/python2.7/dist-packages/zim/gui/pageview.py:2416: GtkWarning: Invalid text buffer iterator: either the iterator is uninitialized, or the characters/pixbufs/widgets in the buffer have been modified since the iterator was created.
You must use marks, character numbers, or line numbers to preserve a position across buffer modifications.
You can apply tags and insert marks without invalidating your iterators,
but any mutation that affects 'indexable' buffer contents (contents that can be referred to by character offset)
will invalidate all outstanding iterators
  if not bound.is_end() and bound.forward_to_tag_toggle(None):
Segmentation fault
```
* 10:51 PM: re-launch from the CLI, took 1m43s to rendered GUI

I can maybe deal with this for this week, but I'm on a trip next week and I require stability then, so I may have to downgrade to 0.65 for that...

**Comment by jaap-karssenberg** [comment]
For issue #7 I have tracked down the root cause to the statement `gobject.threads_init()` in `zim/main/__init__.py`. After removing that line I have seen no more segfaults on my linux system. Not sure if your issue has the same root cause, but it is worth trying.

Will check the other warnings that show in your log separately, but expect they are unrelated to the segfaults.

Might indeed be wise to go with 0.65 if you don't want to risk debugging in the middle of your trip. Would appreciate if you could test the fix mentioned above - if confirmed I'll release 0.67 this weekend.

**Comment by vossenjp** [comment]
2017-05-04 Thu:
* 11:45 AM: First crash of the day...
```
/usr/lib/python2.7/dist-packages/zim/gui/pageview.py:2334: GtkWarning: Invalid text buffer iterator: either the iterator is uninitialized, or the characters/pixbufs/widgets in the buffer have been modified since the iterator was created.
You must use marks, character numbers, or line numbers to preserve a position across buffer modifications.
You can apply tags and insert marks without invalidating your iterators,
but any mutation that affects 'indexable' buffer contents (contents that can be referred to by character offset)
will invalidate all outstanding iterators
  while iter.compare(end) == -1:
/usr/lib/python2.7/dist-packages/zim/gui/pageview.py:2335: GtkWarning: Invalid text buffer iterator: either the iterator is uninitialized, or the characters/pixbufs/widgets in the buffer have been modified since the iterator was created.
You must use marks, character numbers, or line numbers to preserve a position across buffer modifications.
You can apply tags and insert marks without invalidating your iterators,
but any mutation that affects 'indexable' buffer contents (contents that can be referred to by character offset)
will invalidate all outstanding iterators
  pixbuf = iter.get_pixbuf()
Segmentation fault
```
* `sudo vi /usr/lib/python2.7/dist-packages/zim/main/__init__.py`
    * `# per BUG16 gobject.threads_init()`
* 11:50 AM: re-launch from the CLI, took 2m3s to rendered GUI, and it is not where I left off
* Killed via CTRL-C
* `rm /tmp/zim-jp/*`
* 12:01 PM: re-launch from the CLI, took <2s to rendered GUI, and it is not where I left off
    * Killing the old socket files that were building up (duh!) seems to have solved the launch speed problem!

As far as I know I am not doing anything with any external programs or attachments, thus #7 doesn't _seem_ relevant, but I hope the patch works anyway.  :-)

Next:
1. I'll report again at the end of today, but if I forget, no news is good news.  I'll also stop including the crashes like above since they do seem to be repeating.
1. I'll keep the patched 0.66 for Friday and see how it goes.
1. Any idea why it keeps not returning to where I left off?  I assume it only saves that state on a graceful shutdown?  Maybe it should save that more often, unless that's too much overhead.  Of course, that's never been a problem until this instability, so once it's stable again this may be moot.

**Comment by vossenjp** [comment]
* 04:14 PM: crashed (Including Traceback because it's a bit different)
```
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/zim/signals.py", line 414, in callback
    self.cb_func(*arg, **kwarg)
  File "/usr/lib/python2.7/dist-packages/zim/gui/widgets.py", line 2027, in update_completion
    self._fill_completion_for_anchor(path, prefix, text)
  File "/usr/lib/python2.7/dist-packages/zim/gui/widgets.py", line 2053, in _fill_completion_for_anchor
    for p in self.notebook.pages.list_pages(path):
  File "/usr/lib/python2.7/dist-packages/zim/notebook/index/pages.py", line 485, in list_pages
    page_id = self._pages.get_page_id(path) # can raise
  File "/usr/lib/python2.7/dist-packages/zim/notebook/index/pages.py", line 310, in get_page_id
    raise IndexNotFoundError, 'Page not found in index: %s' % pagename.name
zim.notebook.index.base.IndexNotFoundError: Page not found in index: http
/usr/lib/python2.7/dist-packages/zim/gui/pageview.py:2150: GtkWarning: Invalid text buffer iterator: either the iterator is uninitialized, or the characters/pixbufs/widgets in the buffer have been modified since the iterator was created.
You must use marks, character numbers, or line numbers to preserve a position across buffer modifications.
You can apply tags and insert marks without invalidating your iterators,
but any mutation that affects 'indexable' buffer contents (contents that can be referred to by character offset)
will invalidate all outstanding iterators
  pixbuf = iter.get_pixbuf()
Segmentation fault
```
* `rm -rf /tmp/zim-jp/* ./.cache/zim/*`
* Then put a copy of `./.cache/zim/notebook-.../state.conf` back
* 04:31 PM: Re-launch from the CLI, took <2s to rendered GUI, but it is not where I left off
    * So it seems `./.cache/zim/notebook-.../state.conf` is not updated often enough?
    * Re-indexed, and the new index is 67K while the old one was 81K
        * This may be as expected, my "Work" node turns over weekly
* 04:33 PM: Crashed again
```
/usr/lib/python2.7/dist-packages/zim/gui/pageview.py:2416: GtkWarning: Invalid text buffer iterator: either the iterator is uninitialized, or the characters/pixbufs/widgets in the buffer have been modified since the iterator was created.
You must use marks, character numbers, or line numbers to preserve a position across buffer modifications.
You can apply tags and insert marks without invalidating your iterators,
but any mutation that affects 'indexable' buffer contents (contents that can be referred to by character offset)
will invalidate all outstanding iterators
  if not bound.is_end() and bound.forward_to_tag_toggle(None):
Segmentation fault
```
* 04:31 PM: Re-launch from the CLI, took <2s to rendered GUI, but it is not where I left off
* 09:50 PM: Running fine so far, with a lot of activity...

**Comment by jaap-karssenberg** [comment]
Looks like the crashes linked to "Invalid text buffer iterator" are different from the once in issue #7. Would love to reproduce those, as I can't find in the current code where these iterators changed.

Any hint from your work flow when these happen?

**Comment by vossenjp** [comment]
I have no clue what the trigger is, as far as I can tell it's random.  I thought for a while it was if I was typing during a save, but now I don't think so.  That's why all the "lost N characters" above, Zim would crash when I wasn't looking and my typing would end up someplace else, then I could sometimes figure out how many characters made it into Zim after an auto-save but before the crash.

It seems _better_ after commenting out that line, and it seems to crash twice in a row, then be stable for hours.  That just happened this morning:
* 2017-05-04 Thu:
    * 04:31 PM: Re-launch from the CLI, took <2s to rendered GUI, but it is not where I left off
    * 09:50 PM: Running fine so far, with a lot of activity...
* 2017-05-05 Fri
    * 11:07 AM: Crashed...
    * 11:08 PM: Re-launch from the CLI, took <2s to rendered GUI, but it is not where I left off
    * 11:11 AM: Crashed again after only a few changes...
    * 11:30 AM: Running fine so far, with a moderate activity...

**Comment by vossenjp** [comment]
2017-05-05 Fri:

* 11:54 AM: Crashed
* 11:55 PM: Re-launch from the CLI, took <2s to rendered GUI, but it is not where I left off
* 02:22 PM: Crashed
* 02:23 PM: Re-launch from the CLI, took <2s to rendered GUI, but it is not where I left off
* 02:46 PM: Crashed
* 02:46 PM: Re-launch from the CLI...
* 02:48 PM: Crashed
* 02:48 PM: Re-launch from the CLI...
* 03:20 PM: Crashed & re-launch from the CLI...
* 03:22 PM: Crashed & re-launch from the CLI...
* While in my "con call" tab trying to take notes:
    * 03:51 PM: Crashed & re-launch from the CLI...
    * 04:05 PM: Crashed & re-launch from the CLI...
    * 04:15 PM: Crashed & re-launch from the CLI...
    * 04:28 PM: Crashed & re-launch from the CLI...
    * 04:29 PM: Crashed & re-launch from the CLI...
    * 04:46 PM: Crashed & re-launch from the CLI...
    * 04:58 PM: Crashed & re-launch from the CLI...
    * 04:59 PM: Crashed & re-launch from the CLI... Turn off Spell checker.
    * 05:17 PM: Crashed & re-launch from the CLI... Turn off Spell checker.
    * 05:21 PM: Crashed & re-launch from the CLI... Turn off Spell checker.
    * 05:23 PM: Crashed & re-launch from the CLI... Turn off Spell checker.
    * 05:27 PM: Crashed & re-launch from the CLI...
    * 05:46 PM: Crashed & re-launch from the CLI...
    * 05:50 PM: Crashed.  `vi ~/.cache/zim/notebook-.../state.conf` and clear some things:
```
WARNING: Invalid config value for list: "" - Value not allowed to be empty
WARNING: Invalid config value for recent: "" - Value not allowed to be empty
WARNING: Invalid config value for current: "" - Value not allowed to be empty
```
* 05:52 PM: Re-launch from the CLI...
* 06:13 PM: Re-launch from the CLI...
    * State file NOT saved, (no surprise...)
    * And I'm **really** paranoid about using "File > Quit" since that has killed my VM twice, which is exceptionally annoying

2017-05-06 Sat:

* 11:56 AM: Crashed as soon as I started using Zim and typing... Re-launch from the CLI...
* 11:57 AM: Crashed while typing, re-launch from the CLI...
```
/usr/lib/python2.7/dist-packages/zim/gui/pageview.py:2191: GtkWarning: Invalid text buffer iterator: either the iterator is uninitialized, or the characters/pixbufs/widgets in the buffer have been modified since the iterator was created.
You must use marks, character numbers, or line numbers to preserve a position across buffer modifications.
You can apply tags and insert marks without invalidating your iterators,
but any mutation that affects 'indexable' buffer contents (contents that can be referred to by character offset)
will invalidate all outstanding iterators
  iter.forward_char()
Segmentation fault
```
* 12:00 PM: Crashed while typing, re-launch from the CLI...
* 02:18 PM: OK, I give up, reverting to 0.65
    * `$ sudo aptitude remove zim`
    * `$ wget 'http://zim-wiki.org/downloads/zim_0.65_all.deb'`
    * `$ sudo dpkg -i ~/working/zim_0.65_all.deb`
* Launch Zim 0.65 from icon
    * CTRL-HOME doesn't work, as expected, but that's an easy fix
        * See https://bugs.launchpad.net/zim/+bug/1513766 and
            * https://launchpadlibrarian.net/229502335/pageview1.diff
            * https://launchpadlibrarian.net/229502375/pageview2.diff
        * `sudo vi /usr/share/pyshared/zim/gui/pageview.py`
            * Patch 3 lines as per diffs above
    * Zim is not rendering page markup!
        * To make a **long** story short, I somehow got a bunch of `'''` markers in a few of my pages!
        * I have no clue at all how that could happen, but the fix was to remove the bad markers
        * One thing I thought was that the package above wasn't quite "right" so I re-installed the exact mostly-working (except for CTRL-HOME) pacakge I had:
            * `$ wget https://launchpad.net/~jaap.karssenberg/+archive/ubuntu/zim/+files/zim_0.65~ubuntu14.04.1_all.deb`
            * `$ sudo dpkg -i /home/jp/working/zim_0.65~ubuntu14.04.1_all.deb`
    * Closing Zim via the `x` in the `- + x` title bar works fine and does not crash my VM anymore...

It has occurred to me that Zim 0.66 has **always** crashed **while I'm typing in it**!  I've never tried to switch to it and found it gone:
* New indexer?
* Turning off the Spell checker didn't help

Zim 0.66 was not remembering where I left off when switching between nodes either!  But 0.65 does remember that state.

**Comment by jaap-karssenberg** [comment]
You are right, given the error message and given the fact that it still is happening, this issue is different from  #7 . I was lef in that direction because your early logs shows "spawning firefox" before crashes, but that seems to be accidental.

I have no experience with VM, but the fact that a crash takes down the whole VM is scary. No ideas on my side how to further debug.

**Comment by jaap-karssenberg** [comment]
P.S. the "noise" items above giving IndexNotFound errors are harmless, checked in patch to hide them

**Comment by jaap-karssenberg** [comment]
The line numbers mentioned in the GtkWarning lines are either in the "get_parsetree()" method or in functions that could be called from this method. This leads to the hypothesis that there is something threading related going on. Could well be that the auto-save function triggers these crashes - that would be consistent with the crashes only happening while typing. Running zim with "-D" would print out more info to troubleshoot.

Also I added a hidden option "`autosave_use_thread=True|False`" in the section "`[GtkInterface]`" of '`preferences.conf`'. This way you can disable thread usage during auto save. Not sure if it helps, but only thing I can think of now.  (See cab93d3a4c5a607fd47d43bb37bae86ba7a33e11)

Regards,

Jaap

**Comment by vossenjp** [comment]
First, I have to admit to a silly reporting error in this bug.  I said "As far as I know I am not doing anything with any external programs" but I failed  consider that clicking on a link that launches FF is "an external program."  Duh.  So all those FF launches are legitimate, I do link to our Redmine issues and do click on those links now and then...  So "spawning firefox" is legit--but--Zim never crashed when I did that.

As for the VM crashes, I'm inclined to blame either Windows or something wonky on my side interacting with whatever is wrong  with Zim.  I do have GUI issues with the VM from time to time, and my WAG is that they are related somehow to GTK, but I could be wrong.  I was running Zim 0.65 on the Windows side rarely, and I totally stopped to see if that would affect the other VM problems and it did not.  That said, 0.65 and older never had any problems inside the VM and certainly never crashed the VM.

Now, threading is an interesting note.  My VM has only 1 virtual CPU because it never seemed to need more.  Maybe something is confused because the CPU reports thread-capable but it really isn't or something is just odd?

Windows (host) side `VBoxManage showvminfo` says:
```
...
Number of CPUs:  1
PAE:             off
Long Mode:       on
...
```
And inside the Mint-17.3 VM `lshw` says:
```
$ sudo lshw -class cpu
  *-cpu
       product: Intel(R) Core(TM) i5-6300U CPU @ 2.40GHz
       vendor: Intel Corp.
       physical id: 2
       bus info: cpu@0
       width: 64 bits
       capabilities: fpu fpu_exception wp vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 syscall nx rdtscp x86-64 constant_tsc rep_good nopl xtopology nonstop_tsc pni pclmulqdq monitor ssse3 cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx rdrand hypervisor lahf_lm abm 3dnowprefetch rdseed clflushopt
```

I'm traveling the rest of this week and so I doubt I'll be able to test the RC version with the new thread setting off, unless I get really bored, which is doubtful.  I _might_ be able to get to this on Friday...

**Comment by jaap-karssenberg** [comment]
> I'm traveling the rest of this week and so I doubt I'll be able to test the RC version with the new thread setting off, unless I get really bored, which is doubtful. I might be able to get to this on Friday...

So far I have not seen other users affected by the same issue, so for now I'm assuming that this is specific to the VM environment. Under that assumption I won't hold up the 0.67 release on this, so don't rush yourself testing. That said of course I want to work to fix is as it is a regression since 0.65, but could end up in release n+1.

Regards,

Jaap

**Comment by vossenjp** [comment]
I didn't want to jinx it by saying anything too soon, but zim-0.67~rc1~ubuntu14.04.1 updated via PPA on 2017-05-13 Sat and I restarted into it on 2017-05-15 Mon and it has been rock solid.  Not a single crash or glitch or anything.  So...wow, I don't know what it was about 0.66 that didn't like me or my VM, but it seems to be gone in 0.67~rc1.
